byron/ledger/executable-spec/test/Ledger/Delegation/Properties.hs

```diff
@@ -259,18 +271,39 @@ instance HasTrace DBLOCK where
         n <- Gen.integral (Range.linear 0 13)
         pure $! Set.fromAscList $ mkVKeyGenesis <$> [0 .. n]
 
-  sigGen _ (env, st) =
-    DBlock <$> nextSlotGen <*> sigGen @DELEG env st
+  sigGen (Just InvalidDelegationCertificate) _ (env, _st) =
+    DBlock <$> nextSlotGen env <*> Gen.list (Range.constant 0 10) randomDCertGen
     where
-      -- We want the resulting trace to include a large number of epoch
-      -- changes, so we generate an epoch change with higher frequency.
-      nextSlotGen =
-        incSlot <$> Gen.frequency
-                      [ (1, Gen.integral (Range.constant 1 10))
-                      , (2, pure $! slotsPerEpoch (_dSEnvK env))
-                      ]
+      -- | Generate a random delegation certificate, which has a high probability of failing since
+      -- we do not consider the current delegation state. So for instance, we could generate a
+      -- delegation certificate for a genesis key that already delegated in this epoch.
+      randomDCertGen :: Gen DCert
+      randomDCertGen = do
+        (vkg, vk, e) <- (,,) <$> vkgGen' <*> vkGen' <*> epochGen'
+        pure $! mkDCert vkg (signWithGenesisKey vkg (vk, e)) vk e
         where
-          incSlot c = (env ^.slot) `addSlot` SlotCount c
+          vkgGen' = Gen.element $ Set.toList allowed
+          allowed = _dSEnvAllowedDelegators env
+          vkGen' = Gen.element $ VKey . Owner <$> [0 .. (2 * fromIntegral (length allowed))]
+          epochGen' =  Epoch
+                    .  fromIntegral -- We don't care about underflow. We want to generate large epochs anyway.
+                    .  (fromIntegral n +)
+                   <$> Gen.integral (Range.constant (-2 :: Int) 2)
```

Comment:
Is this range small for testing purposes?

Reply:
I don't think so, the idea is that we can generate duplicated certificates with a relatively high probability. If we make the range larger we will decrease the probability of collisions. However we could have made the constants a bit larger (like 5 instead of 2).

Reply:
Got it, makes sense.